File: DSA/LeetCode/Backtracking/552-Student-Attendance-Record-II-iterative.js
/**
 * @param {number} n
 * @return {number}
 */
var checkRecord = function (n) {
    const modulus = 1000000007;

    // dp[i][j][k] means the number of valid sequences of length i
    // with j 'A's and ending with 'k' consecutive 'L's
    let dp = Array.from({ length: n + 1 }, () =>
        Array.from({ length: 2 }, () => Array(3).fill(0))
    );

    // Base case: one sequence of length 0 (empty sequence)
    dp[0][0][0] = 1;

    for (let i = 1; i <= n; i++) {
        for (let j = 0; j < 2; j++) {  // Number of 'A's
            for (let k = 0; k < 3; k++) {  // Number of trailing 'L's

                // Adding 'P' to the sequence
                dp[i][j][0] = (dp[i][j][0] + dp[i - 1][j][k]) % modulus;

                // Adding 'A' to the sequence (only if no 'A' has been added yet)
                if (j > 0) {
                    dp[i][j][0] = (dp[i][j][0] + dp[i - 1][j - 1][k]) % modulus;
                }

                // Adding 'L' to the sequence (only if there are less than 2 trailing 'L's)
                if (k > 0) {
                    dp[i][j][k] = (dp[i][j][k] + dp[i - 1][j][k - 1]) % modulus;
                }
            }
        }
    }

    // Sum all valid sequences of length n
    let result = 0;
    for (let j = 0; j < 2; j++) {
        for (let k = 0; k < 3; k++) {
            result = (result + dp[n][j][k]) % modulus;
        }
    }

    return result;
};

File: DSA/LeetCode/Backtracking/552-Student-Attendance-Record-II-iterative.py
class Solution(object):
    def checkRecord(self, n):
        MOD = 10**9 + 7
    
        # Initialize dp table
        dp = [[[0] * 3 for _ in range(2)] for _ in range(n+1)]
        dp[0][0][0] = 1  # Base case
        
        for i in range(n):
            for j in range(2):
                for k in range(3):
                    # If we append 'P'
                    dp[i+1][j][0] = (dp[i+1][j][0] + dp[i][j][k]) % MOD
                    
                    # If we append 'A'
                    if j < 1:
                        dp[i+1][j+1][0] = (dp[i+1][j+1][0] + dp[i][j][k]) % MOD
                    
                    # If we append 'L'
                    if k < 2:
                        dp[i+1][j][k+1] = (dp[i+1][j][k+1] + dp[i][j][k]) % MOD
        
        # Sum up all valid sequences of length n
        result = 0
        for j in range(2):
            for k in range(3):
                result = (result + dp[n][j][k]) % MOD
        
        return result
        